contracts/0.8.25/upgrade/V3Template.sol

```diff
@@ -0,0 +1,464 @@
+*   - `finishUpgrade()` as the last step of the upgrade
+*/
+contract V3Template {
+    //
+    // Events
+    //
+
+    event UpgradeFinished();
+
+    //
+    // Constants and immutables
+    //
+
+    uint256 internal constant ACCOUNTING_ORACLE_CONSENSUS_VERSION = 4;
+    uint256 internal constant EXPECTED_FINAL_LIDO_VERSION = 3;
+    uint256 internal constant EXPECTED_FINAL_ACCOUNTING_ORACLE_VERSION = 3;
+
+    V3Addresses public immutable ADDRESSES;
+
+    //
+    // -------- Roles --------
+    //
+    bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
+    // Burner
+    bytes32 public immutable REQUEST_BURN_SHARES_ROLE;
+    bytes32 public immutable REQUEST_BURN_MY_STETH_ROLE;
+    // PauseUntilWithRoles
+    bytes32 public immutable RESUME_ROLE;
+    bytes32 public immutable PAUSE_ROLE;
+    // OracleReportSanityChecker
+    bytes32 public immutable ALL_LIMITS_MANAGER_ROLE;
+    bytes32 public immutable EXITED_VALIDATORS_PER_DAY_LIMIT_MANAGER_ROLE;
+    bytes32 public immutable APPEARED_VALIDATORS_PER_DAY_LIMIT_MANAGER_ROLE;
+    bytes32 public immutable ANNUAL_BALANCE_INCREASE_LIMIT_MANAGER_ROLE;
+    bytes32 public immutable SHARE_RATE_DEVIATION_LIMIT_MANAGER_ROLE;
+    bytes32 public immutable MAX_VALIDATOR_EXIT_REQUESTS_PER_REPORT_ROLE;
+    bytes32 public immutable MAX_ITEMS_PER_EXTRA_DATA_TRANSACTION_ROLE;
+    bytes32 public immutable MAX_NODE_OPERATORS_PER_EXTRA_DATA_ITEM_ROLE;
+    bytes32 public immutable REQUEST_TIMESTAMP_MARGIN_MANAGER_ROLE;
+    bytes32 public immutable MAX_POSITIVE_TOKEN_REBASE_MANAGER_ROLE;
+    bytes32 public immutable SECOND_OPINION_MANAGER_ROLE;
+    bytes32 public immutable INITIAL_SLASHING_AND_PENALTIES_MANAGER_ROLE;
+    // StakingRouter
+    bytes32 public immutable REPORT_REWARDS_MINTED_ROLE;
+    // VaultHub
+    bytes32 public immutable VAULT_MASTER_ROLE;
+    bytes32 public immutable VAULT_REGISTRY_ROLE;
+    // OperatorGrid
+    bytes32 public immutable REGISTRY_ROLE;
+
+    //
+    // -------- Constants --------
+    //
+
+    // Timestamp since startUpgrade() and finishUpgrade() revert with Expired()
+    // This behavior is introduced to disarm the template if the upgrade voting creation or enactment didn't
+    // happen in proper time period
+    uint256 public constant EXPIRE_SINCE_INCLUSIVE = 1754006400; // 2025-08-01 00:00:00 UTC
+
+    // Initial value of _upgradeBlockNumber
+    uint256 internal constant UPGRADE_NOT_STARTED = 0;
+
+    uint256 internal constant INFINITE_ALLOWANCE = type(uint256).max;
+
+    //
+    // Structured storage
+    //
+
+    uint256 public upgradeBlockNumber = UPGRADE_NOT_STARTED;
+    bool public isUpgradeFinished;
+    uint256 internal initialOldBurnerStethSharesBalance;
+    uint256 internal initialTotalShares;
+    uint256 internal initialTotalPooledEther;
+
+
+    /// @param _addresses V3Addresses contract address
+    constructor(address _addresses) {
+        ADDRESSES = V3Addresses(_addresses);
```

Comment:
Maybe, inherit Template from V3Addresses instead of deploying it as a separate address

Reply:
Yeah, would rather inherit